elastic/kibana
# [ILM ] Fix logic for showing/hiding recommended allocation on Cloud

Relates to logic originally implemented in https://github.com/elastic/kibana/pull/81455

## Summary

On cloud, hide the "X nodes (recommended)" option when `data: true` is selected. Previously we only hid this option when legacy `data: true` was detected AND no new `data_*` roles were present. This PR is limited to fixing this logic and updating tests to hide the "recommended" option when legacy, deprecated `data:true` config is in use on cloud.

<img width="978" alt="Screenshot 2021-02-08 at 12 40 30" src="https://user-images.githubusercontent.com/8155004/107215479-a62bb980-6a0b-11eb-9bbe-e86e4a219c9e.png">

## Requirements matrix

**Please note**: this matrix reflects node types (deprecated `data:true` config) and using node roles without the presence of autoscaling. The case where autoscaling is on still needs to be addressed as a separate piece of work.

|                                          |   **Using deprecated `data:true`**  | **Using node roles and autoscaling disabled** |
|--------------------------|---------------------|-------------------------------------------------|
| **Deployment >= 7.10 and < 8.0** | Users can only use node attributes to manage shard allocation in ILM. Users can migrate to the new node roles *. We can display a CTA to migrate to node roles in case we detect the deployment is still not using the new data tier roles. We can also hide the dropdown for selecting node roles, which accommodates ECE users where node roles might not be supported by the deployment template. | Users can use both data tier roles and node attributes to manage shard allocation in ILM. We can assume all data tiers are included (available to be sized and used) in the template. In case we can detect a certain data tier is unused (size = 0), for example, there are no warm nodes available, we can show a CTA to enable that tier in the cloud console |
| **Deployment >= 8.0** | Not supported | Same as above |

## Screenshots

**Deployment >= 7.10 and < 8.0 AND deprecated config**

We show a callout telling users they can migrate to the data tiers on cloud. This callout will show when a user is using deprecated config and has selected "custom". If they have no node attributes, the missing node attributes callout will be shown instead, but this should never be the case on cloud.

<img width="970" alt="Screenshot 2021-02-09 at 11 43 18" src="https://user-images.githubusercontent.com/8155004/107353155-03d40a80-6acd-11eb-9971-b9621ebbcdae.png">

**Deployment >= 7.10 and < 8.0 AND node role config**

**Warm tier callout**
<img width="981" alt="Screenshot 2021-02-09 at 11 47 17" src="https://user-images.githubusercontent.com/8155004/107353927-e2bfe980-6acd-11eb-8a98-77f7ac430603.png">

**Cold tier callout (different because cold tier has slider)**

<img width="1006" alt="Screenshot 2021-02-09 at 11 47 28" src="https://user-images.githubusercontent.com/8155004/107354029-f8cdaa00-6acd-11eb-8e2b-5c46c51631a8.png">

## How to test

0. First we will test the non-deprecated config on cloud
0.1. You can set node roles using: `yarn es -E node.roles=data_hot,data_content,master`
1. Simulate running on cloud by adding values to `kibana.dev.yml`:

```yml
xpack.cloud.id: 'eastus2.azure.elastic-cloud.com:9243$59ef636c6917463db140321484d63cfa$a8b109c08adc43279ef48f29af1a3911'
xpack.cloud.deploymentUrl: 'https://cloud.elastic.co/deployments/resolve/cluster/eastus2/9243$59ef636c6917463db140321484d63cfa/'
```

2. Navigate to ILM UI in management section
3. Edit or create a policy
4. Enable the "warm" and "cold" phase and check the options listed under the "data allocation" field. Check that the callouts make sense.
5. You can set a node to "data:true" (legacy data role mode) by running es with: `yarn es snapshot -E node.data=true`. Do steps 2-4 again.

### Checklist

- [ ] New copy reviewed
